src/main/scala/shine/DPIA/Compilation/AcceptorTranslation.scala

```diff
@@ -0,0 +1,356 @@
+
+      val otype = AccType(dt2)
+      val o = Identifier(freshName("fede_o"), otype)
+
+      acc(array)(MapAcc(n, dt2, dt1,
+        Lambda(o, fedAcc(scala.Predef.Map((x, o)))(f(x))(λ(otype)(x => x))),
+        A))
+
+    case MapFst(w, dt1, dt2, dt3, f, record) =>
+      val x = Identifier(freshName("fede_x"), ExpType(dt1, write))
+
+      val otype = AccType(dt3)
+      val o = Identifier(freshName("fede_o"), otype)
+
+      acc(record)(MapFstAcc(dt1, dt2, dt3,
+        Lambda(o, fedAcc(scala.Predef.Map(x -> o))(f(x))(fun(otype)(x => x))),
+        A))
+
+    case mapSeq@MapSeq(unroll) =>
+      val (n, dt1, dt2, f, array) = mapSeq.unwrap
+      con(array)(λ(expT(n`.`dt1, read))(x =>
+        MapSeqI(unroll)(n, dt1, dt2,
+          fun(expT(dt1, read))(x =>
+            fun(accT(dt2))(o =>
+              acc(f(x))(o))),
+          x, A)))
+
+    case MapSnd(w, dt1, dt2, dt3, f, record) =>
+      val x = Identifier(freshName("fede_x"), ExpType(dt2, write))
+
+      val otype = AccType(dt3)
+      val o = Identifier(freshName("fede_o"), otype)
+
+      acc(record)(MapSndAcc(dt1, dt2, dt3,
+        Lambda(o, fedAcc(scala.Predef.Map(x -> o))(f(x))(fun(otype)(x => x))),
+        A))
+
+    case MapVec(n, dt1, dt2, f, array) =>
+      con(array)(λ(expT(vec(n, dt1), read))(x =>
+        MapVecI(n, dt1, dt2, λ(expT(dt1, read))(x => λ(accT(dt2))(o => acc(f(x))(o))), x, A)))
+
+    case PadEmpty(n, r, dt, array) =>
+      acc(array)(TakeAcc(n, r, dt, A))
+
+    case PrintType(msg, dt, access, input) =>
+      acc(input)(A)
+
+    case reduceSeq@ReduceSeq(unroll) =>
+      val (n, dt1, dt2, f, init, array) = reduceSeq.unwrap
+      con(reduceSeq)(λ(expT(dt2, write))(r =>
+        acc(r)(A)))
+
+    case Reorder(n, dt, access, idxF, idxFinv, input) =>
+      acc(input)(ReorderAcc(n, dt, idxFinv, A))
+
+    case ScanSeq(n, dt1, dt2, f, init, array) =>
+      con(array)(λ(expT(n`.`dt1, read))(x =>
+        con(init)(λ(expT(dt2, read))(y =>
+          ScanSeqI(n, dt1, dt2,
+            λ(expT(dt1, read))(x => λ(expT(dt2, read))(y => λ(accT(dt2))(o =>
+              acc(f(x)(y))(o)))),
+            y, x, A)))))
+
+    case Scatter(n, m, dt, indices, input) =>
+      con(indices)(fun(expT(m`.`idx(n), read))(y =>
+        acc(input)(ScatterAcc(n, m, dt, y, A))))
+
+    case slide@Slide(n, sz, sp, dt, input) =>
+      con(slide)(λ(expT(n`.`(sz`.`dt), read))(x =>
+        A :=|(n`.`(sz`.`dt))| x ))
+
+    case Split(n, m, w, dt, array) =>
+      acc(array)(SplitAcc(n, m, dt, A))
+
+    case Transpose(n, m, dt, access, array) =>
+      acc(array)(TransposeAcc(n, m, dt, A))
+
+    case Unzip(n, dt1, dt2, access, e) =>
+      acc(e)(UnzipAcc(n, dt1, dt2, A))
+
+    case VectorFromScalar(n, dt, arg) =>
+      con(arg)(λ(expT(dt, read))(e =>
+        A :=|VectorType(n, dt)| VectorFromScalar(n, dt, e)))
+
+    case Zip(n, dt1, dt2, access, e1, e2) =>
+      acc(e1)(ZipAcc1(n, dt1, dt2, A)) `;`
+        acc(e2)(ZipAcc2(n, dt1, dt2, A))
+
+    // OpenMP
```

Comment:
Can we have instead a `shine.OpenMP.Compilation.AcceptorTranslation` which calls `shine.DPIA.Compilation.AcceptorTranslation`? Same for OpenCL and CUDA.

Reply:
I think this is a good idea. I won't have time to do this over the next couple of days though ...